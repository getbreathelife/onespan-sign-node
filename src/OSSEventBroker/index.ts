export * from './OSSEventBroker';
export * from './error';
