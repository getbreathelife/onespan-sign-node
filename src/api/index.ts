import * as RequestBuilders from './requestBuilders';

export { Api } from './Api';
export { RequestBuilders };
