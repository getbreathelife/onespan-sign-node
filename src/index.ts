/**
 * A library to interact with OneSpan Sign's API.
 *
 * @example
 * ```ts
 * const oneSpanSign = new OneSpanSign({
 *   clientId: CLIENT_ID,
 *   secret: CLIENT_SECRET,
 *   type: 'OWNER',   // or 'SENDER'
 * });
 *
 * const { id: packageId } = await oneSpanSign.packages.create({...});
 *
 * await oneSpanSign.documents.create(packageId, ...)
 * ```
 *
 * @packageDocumentation
 */

export { OneSpanSign } from './OneSpanSign';
export * from './api';
export * from './resources';
export * from './types';
export * from './types/requests';
