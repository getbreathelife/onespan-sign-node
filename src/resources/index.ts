export { Resource } from './Resource';
export { DocumentResource } from './DocumentResource';
export { PackageResource } from './PackageResource';
