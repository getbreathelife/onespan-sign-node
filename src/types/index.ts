// Entity typing reference: https://community.onespan.com/sites/default/files/interactive_api_references/oss_1145.yaml
import * as Events from './events';
import * as Requests from './requests';
import * as Responses from './responses';

export * from './api';
export * from './models';
export * from './utils';
export { Requests, Responses, Events };
