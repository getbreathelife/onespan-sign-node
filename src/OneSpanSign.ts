import FormData from 'form-data';

import { Api } from './api';
import {
  CreatePackageRequestPayload,
  CreatePackageResponsePayload,
  DocumentMetadata,
  UploadDocumentRequestPayload,
} from './types';

export class OneSpanSign {
  constructor(private apiKey: string, private apiUrl: string) {}

  public async createPackage(payload: CreatePackageRequestPayload): Promise<CreatePackageResponsePayload> {
    const response = await Api.post(`${this.apiUrl}/api/packages`)
      .withAuthorizationHeader(`Basic ${this.apiKey}`)
      .withBody(payload)
      .fetch();

    return (await response.json()) as CreatePackageResponsePayload;
  }

  public async uploadDocument(
    packageId: string,
    payload: UploadDocumentRequestPayload,
    documentBody: Buffer | ReadableStream
  ): Promise<DocumentMetadata> {
    const formData = new FormData();
    formData.append('file', documentBody, { filename: payload.name });
    formData.append('payload', payload);

    const response = await Api.post(`${this.apiUrl}/api/packages/${packageId}/documents`)
      .withAuthorizationHeader(`Basic ${this.apiKey}`)
      .withBody(formData)
      .fetch();

    return (await response.json()) as DocumentMetadata;
  }
}
