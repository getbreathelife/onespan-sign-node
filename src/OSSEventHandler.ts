import { EventHandler, Events } from 'types';

/**
 * Helper class to ingest
 * {@link https://community.onespan.com/documentation/onespan-sign/guides/feature-guides/developer/setting-callback-notifications | callback notifications}
 * from OneSpan Sign and call the appropriate handler.
 *
 * @public
 */
export class OSSEventBroker {
  protected _documentSignedHandler?: EventHandler<Events.DocumentSignedEvent>;

  public get documentSignedHandler(): EventHandler<Events.DocumentSignedEvent> | undefined {
    return this._documentSignedHandler;
  }

  public set documentSignedHandler(handler) {
    this._documentSignedHandler = handler;
  }
}
