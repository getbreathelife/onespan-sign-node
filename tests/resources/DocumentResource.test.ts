import { Readable } from 'node:stream';

import { AccessTokenOwnerConfig, Api, DocumentResource, Requests } from '../../src';
import { mockApiHappyPath, MockedApi, restoreMockedApi } from '../setup/mockApi';
import { mockedFetch, mockFetchHappyPath } from '../setup/mockNodeFetch';

const MOCK_TOKEN_CONFIG: AccessTokenOwnerConfig = {
  type: 'OWNER',
  clientId: 'clientId',
  secret: 'secret',
};
const MOCK_API_URL = 'http://demo.com';

describe('DocumentResource', () => {
  let mockedApi: MockedApi;
  let documents: DocumentResource;

  beforeAll(() => {
    mockFetchHappyPath();
    mockedApi = mockApiHappyPath();
    documents = new DocumentResource(new Api(MOCK_TOKEN_CONFIG, MOCK_API_URL));
  });

  afterAll(() => {
    restoreMockedApi(mockedApi);
    mockedFetch.mockReset();
  });

  afterEach(() => {
    mockedFetch.mockClear();
  });

  describe('create', () => {
    it.each`
      type        | data                        | expectMatcher
      ${'buffer'} | ${Buffer.from('abc1234')}   | ${(data: any) => data}
      ${'stream'} | ${Readable.from('abc1234')} | ${(data: any) => expect.objectContaining({ source: data })}
    `('sends the provided payload and $type as request body', async ({ data, expectMatcher }) => {
      const payload: Requests.CreateDocumentData = {
        name: 'mock document',
      };
      const packageId = 'mockId';

      await documents.create(packageId, payload, { body: data, filename: 'file.pdf' });

      expect(mockedFetch.mock.calls[0][0]).toMatchSnapshot();

      // Had to compare the value this way instead of using a snapshot because the boundary value is different each time
      expect(mockedFetch.mock.calls[0][1]).toEqual({
        headers: {
          accept: 'application/json',
          authorization: expect.stringMatching(/^Bearer/),
          'content-type': expect.stringMatching(/^multipart\/form-data; boundary=.+/),
        },
        method: 'POST',
        body: expect.objectContaining({
          // eslint-disable-next-line @typescript-eslint/naming-convention
          _streams: expect.arrayContaining([
            expect.stringContaining(`Content-Disposition: form-data; name=\"file\"; filename=\"file.pdf\"`),
            expectMatcher(data),
            expect.stringContaining(`Content-Disposition: form-data; name=\"payload\"`),
            JSON.stringify(payload),
          ]),
        }),
      });
    });
  });

  describe('bulkCreate', () => {
    it('converts the parameters into proper request body', async () => {
      const payload: Requests.BulkCreateDocumentData[] = [
        {
          name: 'document 1',
          documentBody: { body: Buffer.from('abc1234'), filename: 'file1.pdf' },
        },
        {
          name: 'document 2',
          documentBody: { body: Readable.from('abc1234'), filename: 'file2.pdf' },
        },
      ];
      const packageId = 'mockId';

      await documents.bulkCreate(packageId, payload);

      expect(mockedFetch.mock.calls[0][0]).toMatchInlineSnapshot(`"http://demo.com/api/packages/mockId/documents"`);

      expect(mockedFetch.mock.calls[0][1]).toEqual({
        headers: {
          accept: 'application/json',
          authorization: expect.stringMatching(/^Bearer/),
          'content-type': expect.stringMatching(/^multipart\/form-data; boundary=.+/),
        },
        method: 'POST',
        body: expect.objectContaining({
          // eslint-disable-next-line @typescript-eslint/naming-convention
          _streams: expect.arrayContaining([
            expect.stringContaining(`Content-Disposition: form-data; name=\"file\"; filename=\"file1.pdf\"`),
            payload[0].documentBody.body,
            expect.stringContaining(`Content-Disposition: form-data; name=\"file\"; filename=\"file2.pdf\"`),
            expect.objectContaining({ source: payload[1].documentBody.body }),
            expect.stringContaining(`Content-Disposition: form-data; name=\"payload\"`),
            JSON.stringify([{ name: 'document 1' }, { name: 'document 2' }]),
          ]),
        }),
      });
    });
  });

  describe('getMetadata', () => {
    it('sends the request to the correct URL', async () => {
      await documents.getMetadata('package-id', 'documentId');

      expect(mockedFetch.mock.calls[0]).toMatchInlineSnapshot(`
        Array [
          "http://demo.com/api/packages/package-id/documents/documentId",
          Object {
            "headers": Object {
              "accept": "application/json",
              "authorization": "Bearer mockedToken",
              "content-type": "application/json",
            },
            "method": "GET",
          },
        ]
      `);
    });
  });

  describe('getBody', () => {
    it('sends the request to fetch original document correctly', async () => {
      await documents.getBody('package-id', 'documentId', { format: 'original' });

      expect(mockedFetch.mock.calls[0]).toMatchInlineSnapshot(`
        Array [
          "http://demo.com/api/packages/package-id/documents/original",
          Object {
            "headers": Object {
              "accept": "application/octet-stream, application/json",
              "authorization": "Bearer mockedToken",
              "content-type": "application/json",
            },
            "method": "GET",
          },
        ]
      `);
    });

    it('sends the request to fetch PDF document correctly', async () => {
      await documents.getBody('package-id', 'documentId', { format: 'pdf', flatten: true });

      expect(mockedFetch.mock.calls[0]).toMatchInlineSnapshot(`
        Array [
          "http://demo.com/api/packages/package-id/documents/pdf?flatten=true",
          Object {
            "headers": Object {
              "accept": "application/octet-stream, application/json",
              "authorization": "Bearer mockedToken",
              "content-type": "application/json",
            },
            "method": "GET",
          },
        ]
      `);

      await documents.getBody('package-id', 'documentId', { format: 'pdf', flatten: false });

      expect(mockedFetch.mock.calls[1]).toMatchInlineSnapshot(`
        Array [
          "http://demo.com/api/packages/package-id/documents/pdf?flatten=false",
          Object {
            "headers": Object {
              "accept": "application/octet-stream, application/json",
              "authorization": "Bearer mockedToken",
              "content-type": "application/json",
            },
            "method": "GET",
          },
        ]
      `);

      await documents.getBody('package-id', 'documentId', { format: 'pdf' });

      expect(mockedFetch.mock.calls[2]).toMatchInlineSnapshot(`
        Array [
          "http://demo.com/api/packages/package-id/documents/pdf",
          Object {
            "headers": Object {
              "accept": "application/octet-stream, application/json",
              "authorization": "Bearer mockedToken",
              "content-type": "application/json",
            },
            "method": "GET",
          },
        ]
      `);
    });
  });
});
