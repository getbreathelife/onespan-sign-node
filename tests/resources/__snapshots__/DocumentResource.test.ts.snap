// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentResource create sends the provided payload and buffer as request body 1`] = `"http://demo.com/api/packages/mockId/documents"`;

exports[`DocumentResource create sends the provided payload and stream as request body 1`] = `"http://demo.com/api/packages/mockId/documents"`;
